File: Data Structures & Algorithms/invert-a-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# class Solution:
#     def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
#         if not root:
#             return None
        
#         stack = [root]
#         while stack:
#             cur = stack.pop()
#             cur.left, cur.right = cur.right, cur.left
#             if cur.left:
#                 stack.append(cur.left)
            
#             if cur.right:
#                 stack.append(cur.right)
            
#         return root

class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root:
            return None
        
        que = collections.deque([root])
        while que:
            size = len(que)
            for _ in range(size):
                cur = que.popleft()

                cur.left, cur.right = cur.right, cur.left
                if cur.left:
                    que.append(cur.left)
                
                if cur.right:
                    que.append(cur.right)
        return root

File: Data Structures & Algorithms/depth-of-binary-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int maxDepth(TreeNode root) {
        if (root == null) return 0;
        Deque<TreeNode> que = new ArrayDeque<>();
        que.offer(root);
        int depth = 0;
        while (!que.isEmpty()) {
            int size = que.size();
            for (int i = 0; i < size; i++) {
                TreeNode cur = que.poll();
                if (cur.left != null) que.offer(cur.left);
                if (cur.right != null) que.offer(cur.right);
            }
            depth++;
        }
        return depth;
    }
}
